File: src/game/ts/chain_reaction.ts
import { assert, array, CancelPromise } from "@common/util";
import { Hooks } from "@common/hooks";
import { countingArray } from "./util";

export enum CellType {
	Empty,
	Owned,
}

export interface XY {
	x: number;
	y: number;
}

export interface ChainReactionOptions {
	width: number;
	height: number;
	players: number;
}

type EmptyCell = { type: CellType.Empty };
export type OwnedCell = { type: CellType.Owned; owner: number; count: number };
export type Cell = EmptyCell | OwnedCell;

/** Contains common functions for working with `Cell` objects. */
const Cell = {
	/** Mutates a `Cell` into an `EmptyCell`. */
	toEmpty(cell: Cell): EmptyCell {
		const emptyCell = (cell as unknown) as EmptyCell;
		emptyCell.type = CellType.Empty;
		return emptyCell;
	},

	/** Mutates a `Cell` into an `OwnedCell`. */
	toOwned(cell: Cell, owner: number, count: number): OwnedCell {
		const ownedCell = (cell as unknown) as OwnedCell;
		ownedCell.type = CellType.Owned;
		ownedCell.owner = owner;
		ownedCell.count = count;
		return ownedCell;
	},

	/** Creates an `EmptyCell`. */
	empty(): EmptyCell {
		return { type: CellType.Empty };
	},
};

/**
 * Creates a matrix from the provided `cellMatrix`. Each field in the
 * neighbor matrix contains an array of the cells surrounding it. Since
 * getting a cell's neighbors is a common operation in the game, instead
 * of recalculating the neighbors of the same cells every time they're
 * needed, a matrix like this is generated ahead of time.
 *
 * Important Note: To prevent neighbor matrix invalidation, cell
 * references should be preserved. This means that modifying a cell should
 * be done exclusively in a *mutable* fashion.
 */
function neighborMatrix(cellMatrix: Cell[], width: number, height: number) {
	const getXY = (pos: number): XY => ({
		x: pos % width,
		y: Math.floor(pos / width),
	});

	const getPos = ({ x, y }: XY) => y * width + x;

	const neighbors = ({ x, y }: XY): XY[] => [
		{ x: x - 1, y },
		{ x: x + 1, y },
		{ x, y: y - 1 },
		{ x, y: y + 1 },
	];

	const inBounds = ({ x, y }: XY) =>
		x >= 0 && x < width && y >= 0 && y < height;

	const matrix = countingArray(width * height)
		.map(getXY)
		.map(neighbors)
		.map(neighbors => neighbors.filter(inBounds))
		.map(neighbors => neighbors.map(getPos))
		.map(neighbors => neighbors.map(pos => ({ pos, cell: cellMatrix[pos] })));

	return matrix;
}

export class ChainReaction {
	/** Width of the board. */
	readonly width: number;

	/** Height of the board. */
	readonly height: number;

	/** Number of players. */
	readonly players: number;

	/** The game board. */
	grid: Cell[];

	/** Set of positions that have empty cells. */
	emptyCells: Set<number>;

	/** Contains sets of owned positions for each player. */
	ownedCells: Set<number>[];

	/** The next player to play. */
	private _currentPlayer = 0;

	/** Number of players with cells on the game board. */
	private alivePlayers = 0;

	/** Holds scores for all players. */
	private playerScore: number[];

	/** Number of moves made so far. */
	private turn = 0;

	/** Neighbor matrix. See `neighborMatrix()`. */
	private neighbors: Array<Array<{ pos: number; cell: Cell }>>;

	/** Collection of event handlers. */
	hooks = Hooks("explosionDelay", "update");

	private cancelPromise = CancelPromise();

	constructor(options: ChainReactionOptions) {
		const { width, height, players } = options;

		this.width = width;
		this.height = height;
		this.players = players;

		this.grid = array(width * height, Cell.empty);
		this.neighbors = neighborMatrix(this.grid, width, height);
		this.playerScore = array(players, () => 0);
		this.emptyCells = new Set(countingArray(width * height));
		this.ownedCells = array(players, () => new Set());
	}

	get currentPlayer(): number {
		return this._currentPlayer;
	}

	/** Converts a 2D position into a 1D position. */
	private getPos(x: number, y: number) {
		return y * this.width + x;
	}

	/** Tells whether the cell at the given position should explode. */
	shouldExplode(pos: number): boolean {
		const cell = this.grid[pos];
		return cell.type === CellType.Empty
			? true
			: cell.count >= this.neighbors[pos].length;
	}

	cancel(): void {
		this.cancelPromise.cancel();
		this.hooks.clear();
	}

	/** Tells whether the game is finished. */
	isActive(): boolean {
		// Let each player play one turn. After that, the game is finished
		// when only a single player has cells on the board.
		return this.turn <= this.alivePlayers ? true : this.alivePlayers > 1;
	}

	/** Returns the winner if the game is finished. */
	winner(): number {
		if (this.isActive()) {
			throw new Error("Game is still active.");
		}

		return this.playerScore.findIndex(x => x > 0);
	}

	/** Mutates the game object into its original, unaltered state. */
	async reset(): Promise<void> {
		this._currentPlayer = 0;
		this.turn = 0;
		this.alivePlayers = 0;
		this.playerScore = this.playerScore.map(() => 0);
		this.grid.forEach(Cell.toEmpty);
		this.emptyCells.clear();
		countingArray(this.width * this.height).forEach(x =>
			this.emptyCells.add(x),
		);
		this.ownedCells.forEach(set => set.clear());
		await this.hooks.run("update", this.cancelPromise.promise);
	}

	/** Tells whether a cell can be placed at the given position. */
	canPlace(x: number, y: number): boolean {
		const pos = this.getPos(x, y);
		const cell = this.grid[pos];
		if (!cell) {
			return false;
		}

		if (cell.type === CellType.Owned) {
			return cell.owner === this.currentPlayer;
		}

		return true;
	}

	/** Updates a player's score while also keeping track of whether the
	 *  player in question has any remaining alive cells.
	 */
	private updateScore(player: number, delta: number) {
		const oldScore = this.playerScore[player];
		this.playerScore[player] += delta;
		const newScore = this.playerScore[player];

		assert(newScore >= 0, `Player ${player} cannot have a negative score.`);

		if (!oldScore && newScore) {
			this.alivePlayers += 1;
		}

		if (oldScore && !newScore) {
			this.alivePlayers -= 1;
		}
	}

	/** Attempts to place a cell at the given position. */
	async place(x: number, y: number): Promise<void> {
		if (this.cancelPromise.cancelled) {
			await this.reset();
			return;
		}

		const pos = this.getPos(x, y);
		const cell = this.grid[pos];

		if (!this.isActive()) {
			throw new Error("Cannot play once game is over.");
		}

		if (cell === undefined) {
			throw new Error(`Field (${x}, ${y}) is outside of bounds.`);
		}

		if (cell.type === CellType.Empty) {
			Cell.toOwned(cell, this._currentPlayer, 1);
			this.ownedCells[this._currentPlayer].add(pos);
			this.emptyCells.delete(pos);
		} else {
			if (cell.owner !== this._currentPlayer) {
				throw new Error(`Field (${x}, ${y}) is already taken.`);
			}

			cell.count += 1;
		}

		assert(cell.type === CellType.Owned);
		this.updateScore(this.currentPlayer, 1);

		await this.hooks.run("update", this.cancelPromise.promise);

		if (this.shouldExplode(pos)) {
			Cell.toEmpty(cell);
			this.emptyCells.add(pos);
			this.ownedCells[this._currentPlayer].delete(pos);
			await this.explode(pos, this._currentPlayer);
		}

		this.nextPlayer();
		this.turn += 1;
	}

	toString(): string {
		let result = "";

		for (let y = 0; y < this.height; y++) {
			for (let x = 0; x < this.width; x++) {
				const pos = this.getPos(x, y);
				const cell = this.grid[pos];
				if (cell.type === CellType.Empty) {
					result += "o";
				} else {
					result += cell.count;
				}

				result += " ";
			}
			result += "\n";
		}

		return result;
	}

	nextPlayer(): void {
		this._currentPlayer = (this._currentPlayer + 1) % this.players;
	}

	/** Given a cell that has reached critical mass, splits
	 *  the cell and takes over the neighboring cells. This
	 *  process then repeats recursively as long as any of
	 *  the neighboring cells also have critical mass.
	 */
	private async explode(origin: number, player: number): Promise<void> {
		let queue = this.neighbors[origin];

		/** Set of cell positions whose cells have reached critical mass.
		 *  It is important that this collection is a set and not an array
		 *  to avoid the following scenario: Imagine a 3x3 game board like this:
		 *  ```
		 *  0 3 0
		 *  3 3 0
		 *  0 0 0
		 *  ```
		 *  The two cells at (0, 1) and (1, 0) explode and each send one mass to
		 *  the cell in the very center of the board. As explosions are processed
		 *  sequentially, the center cell first has 4, then 5 mass. Since the cell
		 *  had reached critical mass already at 4, pushing to an array causes the
		 *  cell to explode twice.
		 *
		 *  Using a set solves the issue by forcing cells to explode only once.
		 */
		const criticals = new Set<number>();

		while (queue.length) {
			const newQueue: typeof queue = [];

			for (const { pos, cell } of queue) {
				this.ownedCells[player].add(pos);

				if (cell.type === CellType.Owned) {
					if (cell.owner !== player) {
						this.ownedCells[cell.owner].delete(pos);
						this.updateScore(player, cell.count);
						this.updateScore(cell.owner, -cell.count);
						cell.owner = player;
					}

					cell.count += 1;
				} else {
					Cell.toOwned(cell, player, 1);
					this.emptyCells.delete(pos);
				}

				assert(cell.type === CellType.Owned);

				if (this.shouldExplode(pos) && this.isActive()) {
					criticals.add(pos);
				}
			}

			// The update hook runs before any of the cells are removed,
			// so that the user has a chance to see which cells had critical
			// mass before they exploded.
			await this.hooks.run("update", this.cancelPromise.promise);
			await this.hooks.run("explosionDelay", this.cancelPromise.promise);

			for (const pos of criticals) {
				newQueue.push(...this.neighbors[pos]);
				const cell = this.grid[pos];
				assert(cell.type === CellType.Owned);

				// Make the cell an empty one only if it has no leftover mass.
				cell.count -= this.neighbors[pos].length;
				if (cell.count === 0) {
					this.ownedCells[cell.owner].delete(pos);
					this.emptyCells.add(pos);
					Cell.toEmpty(cell);
				}
			}

			queue = newQueue;
			criticals.clear();

			if (this.cancelPromise.cancelled) {
				await this.reset();
				break;
			}
		}
	}
}

File: src/game/rust/src/chain_reaction.rs
use crate::array2d::{Array2D, Pos};

use std::{
    cell::{Cell, RefCell},
    collections::HashSet,
    convert::TryInto,
    fmt,
    iter::FromIterator,
};

#[derive(Clone, Copy)]
pub struct FieldData {
    pub owner: usize,
    pub count: u32,
}

#[derive(Clone, Copy)]
pub enum Field {
    Owned(FieldData),
    Empty,
}

impl fmt::Display for Field {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        match self {
            Field::Owned(data) => write!(f, "{}", data.count),
            Field::Empty => write!(f, "o"),
        }
    }
}

pub struct ChainReaction {
    width: usize,
    height: usize,
    players: usize,
    current_player: Cell<usize>,
    move_count: Cell<u32>,

    pub empty_cells: RefCell<HashSet<Pos>>,
    pub owned_cells: RefCell<Vec<HashSet<Pos>>>,

    pub grid: Array2D<Cell<Field>>,
    pub capacity: Array2D<u32>,
    player_count: Vec<Cell<u32>>,
}

impl ChainReaction {
    pub fn new(width: usize, height: usize, players: usize) -> Result<ChainReaction, &'static str> {
        if width == 0 {
            Err("Width cannot be zero.")
        } else if height == 0 {
            Err("Height cannot be zero.")
        } else if players == 0 {
            Err("Player count cannot be zero.")
        } else {
            let grid = Array2D::new(Cell::new(Field::Empty), Pos(width, height));
            let capacity = ChainReaction::get_capacity(width, height);

            let empty_cells_iter = (0..width * height).map(|p| Pos(p % width, p / width));

            Ok(ChainReaction {
                width,
                height,
                players,
                grid,
                capacity,
                current_player: Cell::new(0),
                move_count: Cell::new(0),
                player_count: vec![Cell::new(0); players],

                empty_cells: RefCell::new(HashSet::from_iter(empty_cells_iter)),
                owned_cells: RefCell::new(vec![HashSet::new(); players]),
            })
        }
    }

    pub fn width(&self) -> usize {
        self.width
    }

    pub fn height(&self) -> usize {
        self.height
    }

    pub fn current_player(&self) -> usize {
        self.current_player.get()
    }

    fn change_owner(&self, pos: &Pos, prev_owner: Option<usize>, new_owner: Option<usize>) {
        match (prev_owner, new_owner) {
            (None, Some(new)) => {
                self.empty_cells.borrow_mut().remove(pos);
                self.owned_cells
                    .borrow_mut()
                    .get_mut(new)
                    .map(|x| x.insert(*pos));
            }
            (Some(old), None) => {
                self.owned_cells
                    .borrow_mut()
                    .get_mut(old)
                    .map(|x| x.remove(pos));
                self.empty_cells.borrow_mut().insert(*pos);
            }
            (Some(old), Some(new)) => {
                self.owned_cells
                    .borrow_mut()
                    .get_mut(old)
                    .map(|x| x.remove(pos));
                self.owned_cells
                    .borrow_mut()
                    .get_mut(new)
                    .map(|x| x.insert(*pos));
            }
            (None, None) => {}
        };
    }

    pub fn place(&self, pos: &Pos) -> Result<(), &'static str> {
        let player = self.current_player.get();
        if !self.active() {
            return Err("Cannot play after the game is finished.");
        }

        if player >= self.players {
            return Err("Invalid player.");
        }

        let field = self.grid.get(pos)?;

        let new_data = match field.get() {
            Field::Owned(FieldData { owner, count }) => {
                if owner != player {
                    return Err("Field is already taken.");
                } else {
                    self.change_owner(pos, Some(owner), Some(player));

                    FieldData {
                        owner: player,
                        count: count + 1,
                    }
                }
            }
            Field::Empty => {
                self.change_owner(pos, None, Some(player));

                FieldData {
                    owner: player,
                    count: 1,
                }
            }
        };

        let &capacity = self.capacity.get(pos)?;
        self.increase_player_count(player, 1);
        self.move_count.set(self.move_count.get() + 1);

        if new_data.count >= capacity {
            field.set(Field::Empty);
            self.change_owner(pos, Some(player), None);

            self.explode(pos, player)?;
        } else {
            field.set(Field::Owned(new_data));
        }

        self.current_player.set((player + 1) % self.players);

        Ok(())
    }

    pub fn can_play(&self, pos: &Pos) -> Result<bool, &'static str> {
        let field = self.grid.get(pos)?.get();

        let result = match field {
            Field::Empty => true,
            Field::Owned(FieldData { owner, .. }) => owner == self.current_player.get(),
        };

        Ok(result)
    }

    fn get_capacity(width: usize, height: usize) -> Array2D<u32> {
        let mut grid = Array2D::new(4, Pos(width, height));

        for y in 0..height {
            if let Ok(f) = grid.get_mut(&Pos(0, y)) {
                *f -= 1;
            }

            if let Ok(f) = grid.get_mut(&Pos(width - 1, y)) {
                *f -= 1;
            }
        }

        for x in 0..width {
            if let Ok(f) = grid.get_mut(&Pos(x, 0)) {
                *f -= 1;
            }

            if let Ok(f) = grid.get_mut(&Pos(x, height - 1)) {
                *f -= 1;
            }
        }

        grid
    }

    pub fn neighbors(&self, pos: &Pos) -> Vec<(Pos, &Cell<Field>)> {
        let Pos(x, y) = *pos;

        let offsets = (0..4).map(|i| match i {
            0 => x.checked_sub(1).map(|x| Pos(x, y)),
            1 => Some(Pos(x + 1, y)),
            2 => y.checked_sub(1).map(|y| Pos(x, y)),
            3 => Some(Pos(x, y + 1)),
            _ => unreachable!(),
        });

        let positions: Vec<_> = offsets.flatten().filter(|x| self.in_bounds(x)).collect();

        let cells: Vec<_> = positions
            .iter()
            .map(|pos| self.grid.get(pos))
            .flatten()
            .collect();

        positions.iter().cloned().zip(cells).collect()
    }

    fn in_bounds(&self, &Pos(x, y): &Pos) -> bool {
        x < self.width && y < self.height
    }

    fn increase_player_count(&self, player: usize, delta: u32) {
        self.player_count[player].set(self.player_count[player].get() + delta);
    }

    fn decrease_player_count(&self, player: usize, delta: u32) {
        self.player_count[player].set(self.player_count[player].get() - delta);
    }

    pub fn active(&self) -> bool {
        let alive_players_count: u32 = self
            .player_count
            .iter()
            .filter(|x| x.get() > 0)
            .count()
            .try_into()
            .unwrap();

        if self.move_count.get() <= alive_players_count {
            true
        } else {
            assert!(alive_players_count > 0);
            alive_players_count > 1
        }
    }

    pub fn winner(&self) -> Result<usize, &'static str> {
        if self.active() {
            Err("Game is still in progress.")
        } else {
            self.player_count
                .iter()
                .position(|x| x.get() > 0)
                .ok_or("There is no winner.")
        }
    }

    fn explode(&self, origin: &Pos, player: usize) -> Result<(), &'static str> {
        let mut queue = self.neighbors(origin);

        let mut increase_count = 0;

        while queue.len() > 0 {
            let mut new_queue: Vec<(Pos, &Cell<Field>)> = Vec::new();

            for (pos, field) in &queue {
                let new_data = match field.get() {
                    Field::Owned(data) => {
                        if data.owner != player {
                            self.change_owner(pos, Some(data.owner), Some(player));
                            increase_count += data.count;
                            self.decrease_player_count(data.owner, data.count);
                        }

                        FieldData {
                            owner: player,
                            count: data.count + 1,
                        }
                    }
                    Field::Empty => {
                        self.change_owner(pos, None, Some(player));

                        FieldData {
                            owner: player,
                            count: 1,
                        }
                    }
                };
                let &capacity = self.capacity.get(pos)?;
                if new_data.count >= capacity {
                    field.set(Field::Empty);
                    self.change_owner(pos, Some(player), None);

                    for (neighbor_pos, neighbor) in self.neighbors(pos) {
                        new_queue.push((neighbor_pos, neighbor));
                    }
                } else {
                    field.set(Field::Owned(new_data));
                }
            }

            if self.active() {
                queue = new_queue;
            } else {
                queue = vec![];
            }
        }

        self.increase_player_count(player, increase_count);

        Ok(())
    }
}

impl fmt::Display for ChainReaction {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        self.grid.fmt(f)?;
        write!(f, "\n")?;
        let values: Vec<_> = self.player_count.iter().map(|x| x.get()).collect();
        write!(f, "{:?}", values)?;
        Ok(())
    }
}
